lib/Transforms/Scalar/DCE.cpp

```diff
@@ -120,18 +120,16 @@ bool DCE::runOnFunction(Function &F) {
       //
       for (User::op_iterator OI = I->op_begin(), E = I->op_end(); OI != E; ++OI)
         if (Instruction *Used = dyn_cast<Instruction>(*OI))
-          WorkList.push_back(Used);
+          WorkList.insert(Used);
 
       // Remove the instruction.
       I->eraseFromParent();
 
-      // Remove the instruction from the worklist if it still exists in it.
-      WorkList.erase(std::remove(WorkList.begin(), WorkList.end(), I),
-                     WorkList.end());
-
       MadeChange = true;
       ++DCEEliminated;
     }
+
+	WorkList.pop_back();
```

Comment:
In the event that additional instructions were added to the ```Worklist``` at line 123, I fear this would pop the last of those instead of the ```Instruction``` retrieved at line 114. It's possible that accounts for at least part of the perf win sadly.

If it moved here for fear of the additional cleanup that used to be on lines 124 - 131, that was because ```std::vector``` doesn't prevent duplicates.

Reply:
Oh shoot, you're definitely right here. The move was because if it was popped at the start, then the set insertion wouldn't stop us from adding it again in line 123. Could probably just replace this pop_back with an erase(I) - in fact I think I had that before I switched from unordered_map.